sky/spot/controller.py

```diff
@@ -39,145 +40,73 @@ def __init__(self, job_id: int, task_yaml: str,
-                                                        get_end_time=True)
-                # The job is done.
-                spot_state.set_succeeded(self._job_id, end_time=end_time)
-                break
-
-            if job_status == job_lib.JobStatus.FAILED:
-                # Check the status of the spot cluster. If it is not UP,
-                # the cluster is preempted.
-                (cluster_status,
-                 handle) = backend_utils.refresh_cluster_status_handle(
-                     self._cluster_name, force_refresh=True)
-                if cluster_status == global_user_state.ClusterStatus.UP:
-                    # The user code has probably crashed.
-                    end_time = spot_utils.get_job_timestamp(self.backend,
-                                                            self._cluster_name,
-                                                            get_end_time=True)
-                    logger.info(
-                        'The user job failed. Please check the logs below.\n'
-                        f'== Logs of the user job (ID: {self._job_id}) ==\n')
-                    self.backend.tail_logs(handle,
-                                           None,
-                                           spot_job_id=self._job_id)
-                    logger.info(f'\n== End of logs (ID: {self._job_id}) ==')
-                    spot_state.set_failed(
-                        self._job_id,
-                        failure_type=spot_state.SpotStatus.FAILED,
-                        end_time=end_time)
-                    break
-            # cluster can be down, INIT or STOPPED, based on the interruption
-            # behavior of the cloud.
-            # Failed to connect to the cluster or the cluster is partially down.
-            # job_status is None or job_status == job_lib.JobStatus.FAILED
-            logger.info('The cluster is preempted.')
-            spot_state.set_recovering(self._job_id)
-            recovered_time = self._strategy_executor.recover()
-            spot_state.set_recovered(self._job_id,
-                                     recovered_time=recovered_time)
+        self.cluster_name = spot_utils.generate_spot_cluster_name(
+            self.task_name, self.job_id)
+        self.strategy_executor = recovery_strategy.StrategyExecutor.make(
+            self.cluster_name, self.backend, self.task, retry_until_up)
 
     def start(self):
         """Start the controller."""
         try:
-            self._run()
+            self._handle_signal()
+            controller_task = _controller_run.remote(self)
+            # Signal can interrupt the underlying controller process.
+            ready, _ = ray.wait([controller_task], timeout=0)
+            while not ready:
+                try:
+                    self._handle_signal()
+                except exceptions.SpotUserCancelledError as e:
+                    logger.info('Cancelling...')
```

Comment:
If it makes sense, it will be good to add more to log message (e.g., `Cancelling job {self.job_id}...`)